### PR TITLE
composefs: enable it at runtime

### PR DESCRIPTION
**Pull request by giuseppe**
drop the build tag for building composefs and convert it to an option in `storage.conf`

### CODE REVIEW

File: pkg/config/config_test.go [hunk 1]
```diff
@@ -251,6 +251,15 @@ func TestOverlayOptions(t *testing.T) {
 		t.Fatalf("Expected to find 'skip_mount_home' options, got %v", doptions)
 	}
 
+	options.Overlay.UseComposefs = "true"
+	doptions = GetGraphDriverOptions("overlay", options)
+	if len(doptions) == 0 {
```
Comment (eriksjolund) [inline pkg/config/config_test.go:256]:
```suggestion
	if len(doptions) != 0 {
```

**Comment by cgwalters** [comment]
So I was trying this out like this:

```
$ podman --log-level debug --storage-opt use_composefs=true --root /var/tmp/containers --runroot /run/tmp/containers run --rm -ti quay.io/fedora/fedora:39 echo hello
```

I see

```
DEBU[0000] overlay: use_composefs=true 
DEBU[0000] Cached value indicated that overlay is supported 
DEBU[0000] Cached value indicated that data-only layers for overlay are supported 
DEBU[0000] Cached value indicated that overlay is supported 
DEBU[0000] Cached value indicated that metacopy is not being used 
```
Yet AFAICS `generateComposeFsBlob()` is never being invoked because `ApplyDiffFromStagingDirectory` is never being invoked?  But I don't understand the conditions under which that is invoked.

**Comment by giuseppe** [comment]
> ```
> 
> ```
> 
> 
>     
>   
> 
> Yet AFAICS `generateComposeFsBlob()` is never being invoked because `ApplyDiffFromStagingDirectory` is never being invoked?  But I don't understand the conditions under which that is invoked.

it works only with the zstd:chunked code path now, you need to add this snippet to `storage.conf`:

```
[storage.options]

pull_options = {convert_images = "true", enable_partial_images = "true", use_hard_links = "false", ostree_repos=""}

```

**Comment by cgwalters** [comment]
Ah, OK only operating on zstd:chunked *and* requiring an opt-in *and* previously a build tag definitely makes it hard to use :smile: 
Removing one of those things definitely helps.

I think I've done what you suggested, but I am still not managing to trigger the composefs path.  I do see

```
DEBU[0000] Found image "83a6861c2dbd64301df6dde82f9d99b3659b50478fa9cb7296e21552212a3757" as "83a6861c2dbd64301df6dde82f9d99b3659b50478fa9cb7296e21552212a3757" in local containers storage ([overlay@/var/tmp/containers/storage+/run/tmp/containers:overlay.mountopt=nodev,metacopy=on,overlay.use_composefs=true]@83a6861c2dbd64301df6dde82f9d99b3659b50478fa9cb7296e21552212a3757) 
```

which is different, but apparently `ApplyDiffFromStagingDirectory` still isn't being run.

**Comment by giuseppe** [comment]
you already had the image in the storage?  Was it re-pulled once you enabled the storage options?

**Comment by rhatdan** [comment]
LGTM
@cgwalters @eriksjolund @nalind @vrothberg @saschagrunert PTAL

File: drivers/overlay/overlay.go [hunk 2]
```diff
@@ -387,6 +389,22 @@ func Init(home string, options graphdriver.Options) (graphdriver.Driver, error)
 		}
 	}
 
+	if opts.useComposefs {
+		if unshare.IsRootless() {
+			return nil, fmt.Errorf("composefs is not supported in user namespaces")
```
Comment (saschagrunert) [inline drivers/overlay/overlay.go:394]: non blocking nit:
```suggestion
			return nil, errors.New("composefs is not supported in user namespaces")
```

File: drivers/overlay/overlay.go [hunk 3]
```diff
@@ -122,6 +123,7 @@ type Driver struct {
 	supportsDType    bool
 	supportsVolatile *bool
 	usingMetacopy    bool
+	usingComposefs   bool
```
Comment (saschagrunert) [inline drivers/overlay/overlay.go:126]: non-blocking nit:
```suggestion
	useComposefs   bool
```

**Review by saschagrunert — APPROVED** [review]
LGTM

**Comment by rhatdan** [comment]
@giuseppe could you fix @saschagrunert Nit's

**Comment by rhatdan** [comment]
I am going to merge to fix CI/CD system on other PRs.  @giuseppe if you care to fix the nits, please open a new PR.

**Pull request merged**